iothub/device/src/Transport/AmqpIoT/AmqpIoTReceivingLink.cs

```diff
@@ -258,88 +258,104 @@ private void DisposeDelivery(AmqpMessage amqpMessage, bool settled, Accepted acc
+                int status = getStatus(amqpMessage);
 
-                if (!VerifyResponseMessage(amqpMessage))
+                if (status >= 400)
                 {
+                    // Handle failures
                     _onDesiredPropertyReceived.Invoke(null, correlationId, null);
-                }
-
-                Twin twin = null;
-                TwinCollection twinProperties = null;
+                    string error = null;
+                    using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
+                    {
+                        error = reader.ReadToEnd();
+                    }
 
-                if (correlationId != null)
+                    if (status >= 500 || status == 429 || status == 408)
+                    {
+                        throw new IotHubException(error, true);
+                    }
+                    else
+                    {
+                        throw new IotHubException(error, false);
+                    }
+                }
+                else
                 {
-                    if (amqpMessage.BodyStream != null)
+                    Twin twin = null;
+                    TwinCollection twinProperties = null;
+
+                    if (correlationId == null)
                     {
-                        // This a result of a GET TWIN so return (set) the full twin
+                        // Here we are getting desired property update notifications and want to handle it first
+                        using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
+                        {
+                            string patch = reader.ReadToEnd();
+                            twinProperties = JsonConvert.DeserializeObject<TwinCollection>(patch);
+                        }
+                    }
+                    else if (correlationId.StartsWith(AmqpTwinMessageType.Get.ToString(), StringComparison.OrdinalIgnoreCase))
+                    {
+                        // This a response of a GET TWIN so return (set) the full twin
                         using (StreamReader reader = new StreamReader(amqpMessage.BodyStream, System.Text.Encoding.UTF8))
                         {
                             string body = reader.ReadToEnd();
                             var properties = JsonConvert.DeserializeObject<TwinProperties>(body);
                             twin = new Twin(properties);
                         }
                     }
-                    else
+                    else if (correlationId.StartsWith(AmqpTwinMessageType.Patch.ToString(), StringComparison.OrdinalIgnoreCase))
+                    {
+                        // This can be used to coorelate success response with Updating reported properties
+                    }
+                    else if (correlationId.StartsWith(AmqpTwinMessageType.Put.ToString(), StringComparison.OrdinalIgnoreCase))
```

Comment:
we don't seem to be handling the Put scenario inside TwinMessageListener, is this block (the new Twin()) being referenced elsewhere?

Reply:
Because it is not needed for Put.

Reply:
then where is the twin initialization done below used? if we don't reference the twin anywhere, the object initialization in L327 isn't necessary

Reply:
Just fyi: This is not a new code I added here but let me check that.

Reply:
of course, I was pointing out there there might be a scope for us to avoid an unnecessary object initialization if we don't use the value anywhere

Reply:
Got it. 
It is not used apparently and feels safe to remove. Done